src/main/java/com/couchbase/lite/replicator/ChangeTracker.java

```diff
@@ -340,26 +344,40 @@ public void process(HttpRequest request, HttpContext context) throws HttpExcepti
-                                responseOK = receivedPollResponse(fullBody);
+                                try {
+                                    Log.v(Log.TAG_CHANGE_TRACKER, "%s: readValue", this);
+                                    Map<String, Object> fullBody = Manager.getObjectMapper().readValue(inputStream, Map.class);
+                                    Log.v(Log.TAG_CHANGE_TRACKER, "%s: /readValue.  fullBody: %s", this, fullBody);
+                                    responseOK = receivedPollResponse(fullBody);
+                                } catch (JsonParseException jpe) {
+                                    Log.w(Log.TAG_CHANGE_TRACKER, "%s: json parsing error; %s", this, jpe.toString());
+                                }
                             }
                             Log.v(Log.TAG_CHANGE_TRACKER, "%s: responseOK: %s", this, responseOK);
 
-                            if (mode == ChangeTrackerMode.LongPoll && responseOK) {
-
+                            if (responseOK) {
                                 // TODO: this logic is questionable, there's lots
                                 // TODO: of differences in the iOS changetracker code,
                                 client.changeTrackerCaughtUp();
-
                                 Log.v(Log.TAG_CHANGE_TRACKER, "%s: Starting new longpoll", this);
                                 backoff.resetBackoff();
                                 continue;
                             } else {
-                                Log.d(Log.TAG_CHANGE_TRACKER, "%s: Change tracker calling stop (LongPoll)", this);
-                                client.changeTrackerFinished(this);
-                                break;
+                                long elapsed = (System.currentTimeMillis() - startTime) / 1000;
+                                Log.w(Log.TAG_CHANGE_TRACKER, "%s: Longpoll connection closed (by proxy?) after %.1f sec", this, elapsed);
+                                if (elapsed >= 30) {
+                                    // Looks like the connection got closed by a proxy (like AWS' load balancer) while the
+                                    // server was waiting for a change to send, due to lack of activity.
+                                    // Lower the heartbeat time to work around this, and reconnect:
+                                    this.heartBeatSeconds = Math.min(this.heartBeatSeconds, (int) (elapsed * 0.75));
+                                    Log.v(Log.TAG_CHANGE_TRACKER, "%s: Starting new longpoll", this);
+                                    backoff.resetBackoff();
+                                    continue;
+                                } else {
+                                    Log.d(Log.TAG_CHANGE_TRACKER, "%s: Change tracker calling stop (LongPoll)", this);
+                                    client.changeTrackerFinished(this);
+                                    break;
```

Comment:
Ticket for ChangeTracker error handling enhancement
https://github.com/couchbase/couchbase-lite-java-core/issues/639